### ISSUE TITLE
Question about the use of static methods

### ISSUE THREAD

**Issue by potherca** [steos/php-quickcheck#10]
Looking through the code, I encountered _many_ static methods. For various reasons (available on request) this is considered a poor practice for anything that is not a factory(ish) function.

What is the rationale behind making so many things static?

**Comment by steos** [comment]
This is not supposed to be object-oriented code. The fact that the code uses classes at all is just because PHP has such poor support working with standalone functions. Most static methods are actually factory(ish) (e.g. in the Generator class) or utility functions (e.g. the FP class). 

If you have any specific suggestion where the code could be improved in that regard I'm happy to consider it.

**Comment by potherca** [comment]
## Overview

The main problem with making _anyting_ static is that it will sooner or later require _everything_ to be static (class variables, other class methods, etc.).

The main _other_ reason for not wanting static methods is that they are impossible to mock in tests. 

## Testing and mocking

Especially when a "static" method depends on another object, we lose the ability to mock it, making 
the system under test depend on _live_ objects (that can contain bugs of their own) instead of depending on mock/stub objects that will always return the same value/response.

Obviously for "glue" (like factories and dependency injection) and "dumb" input/output logic this is less problematic. But in other places this can already been seen to happen, for instance:
- `Generator::samples` depends on `Random` which can not be mocked when testing it
- `Quick::check` depends on `Random` which can not be mocked when testing it

## Poor support for standalone functions

In regards to PHP having poor support for working with standalone functions, I'm inclined to disagree. Especially when a namespace is utilized this seems no more or less painful than most other languages.

To be fair, I have played around with [custom solutions](https://gist.github.com/potherca/26b52b2618d865ee832d) and [leveraging Composer](https://getcomposer.org/doc/04-schema.md#files) to solve PHP's lack of "autoloading" functions from separate files.

## In conclusion

These are, obviously, "just opinions", but these opinions _are_ based on many years of working in/with _other_ people's code. Not just based on personal preference.

Thank you for your time. :+1:
